File: MyProject/Vetor/Arvore.java
/*
    * Estrutura de dados árvore para armazenar os caracteres
    * Após a contagem, os caracteres com maior frequência são posicionados próximos a raíz da árvore
    * Os de menor frequência são posicionados nas partes inferiores da estrutura
*/

public class Arvore {
    
    private final NoArvore raiz;

    public Arvore(NoArvore raiz) {
        this.raiz = raiz;
    }

    // Retorna a raíz da árvore
    public NoArvore getRaiz() {
        return raiz;
    }

    // Gera a tabela de código, chamando o método private gerarCodigos
    public String[] gerarTabela() {
        String[] tabela = new String[256];
        gerarCodigos(raiz, "", tabela);
        return tabela;
    }
    
    // Gera os binários 0 e 1 no arquivo final
    private void gerarCodigos(NoArvore no, String codigo, String[] tabela) {
        if(no == null) return;

        if(no.ehFolha()) {
            tabela[(int) no.getCaractere()] = codigo.isEmpty() ? "0" : codigo;
            return;
        }

        gerarCodigos(no.getEsquerda(), codigo + "0", tabela);
        gerarCodigos(no.getDireita(), codigo + "1", tabela);
    }

    public void imprimirArvore() {
        imprimirPreOrdem(raiz, true);
    }

    private void imprimirPreOrdem(NoArvore no, boolean ehRaiz) {
        if(no == null) return;

        if(ehRaiz) {
            System.out.println("- (RAIZ, " + no.getFrequencia() + ")");
        } else if(no.ehFolha()) {
            System.out.println("- ('" + no.getCaractere() + "', " + no.getFrequencia() + ")");
        } else {
            System.out.println("- (interno, " + no.getFrequencia() + ")");
        }

        imprimirPreOrdem(no.getEsquerda(), false);
        imprimirPreOrdem(no.getDireita(), false);
    }
}

File: MyProject/C++/Arvore.cpp
#include "Arvore.hpp"
#include <iostream>

Arvore::Arvore(NoArvore* raiz) : raiz(raiz) {}

Arvore::~Arvore() {
    delete raiz;
}

NoArvore* Arvore::getRaiz() const { return raiz; }

void Arvore::gerarTabela(std::string tabela[]) const {
    gerarCodigos(raiz, "", tabela);
}

void Arvore::gerarCodigos(NoArvore* no, const std::string& codigo, std::string tabela[]) const {
    if(no == nullptr) return;

    if(no->ehFolha()) {
        tabela[(unsigned char) no->getCaractere()] = codigo.empty() ? "0" : codigo;
        return;
    }

    gerarCodigos(no->getEsquerda(), codigo + "0", tabela);
    gerarCodigos(no->getDireita(), codigo + "1", tabela);
}

void Arvore::imprimirArvore() const {
    imprimirPreOrdem(raiz, true);
}

void Arvore::imprimirPreOrdem(NoArvore* no, bool ehRaiz) const {
    if(no == nullptr) return;

    if(ehRaiz) {
        std::cout << "- (RAIZ, " << no->getFrequencia() << ")" << std::endl;
    } else if(no->ehFolha()) {
        std::cout << "- ('" << no->getCaractere() << "', " << no->getFrequencia() << ")" << std::endl;
    } else {
        std::cout << "- (interno, " << no->getFrequencia() << ")" << std::endl;
    }

    imprimirPreOrdem(no->getEsquerda(), false);
    imprimirPreOrdem(no->getDireita(), false);
}